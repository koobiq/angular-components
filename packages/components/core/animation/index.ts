export * from './animation';
export { fadeAnimation } from './fade-animations';
export { kbqSelectAnimations } from './select-animations';
