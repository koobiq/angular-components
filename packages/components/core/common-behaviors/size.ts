/**
 * Defines the available size options for Koobiq components.
 */
export type KbqDefaultSizes = 'compact' | 'normal' | 'big';
