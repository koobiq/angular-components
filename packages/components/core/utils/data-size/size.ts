export const formatDataSize = (value: number, precision: number, system): { value: string; unit: string } => {
    const { result, unit } = getHumanizedBytes(value, system);
    let volume: string;

    if (system.abbreviations[0] === unit) {
        volume = result.toString();
    } else {
        volume = result.toFixed(precision).replace(/\./g, ',');
    }

    return {
        value: volume,
        unit
    };
};

/**
 * Переводит байты в Кб, Мб, Гб
 *
 * @param value количество байт
 * @param system система измерения
 * @param threshold нижний порог подсчета
 */
export const getHumanizedBytes = (value: number, system, threshold?: number): { result: number; unit: string } => {
    if (!Number.isFinite(value)) {
        throw new Error('Argument "value" must be number!');
    }

    const calculatedThreshold = Number.isFinite(threshold) ? threshold : Math.pow(system.base, system.power);
    const orderOfMagnitude: number = Math.pow(system.base, system.power);

    let result: number = value;
    let step = 0;
    const len: number = system.abbreviations.length - 1;

    while (step < len) {
        if (calculatedThreshold !== undefined && result < calculatedThreshold) {
            break;
        }

        step++;
        result = value / Math.pow(orderOfMagnitude, step);
    }

    return {
        result,
        unit: system.abbreviations[step]
    };
};
