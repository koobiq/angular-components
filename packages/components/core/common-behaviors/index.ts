import { InjectionToken } from '@angular/core';

export { CanColor, CanColorCtor, KbqComponentColors, ThemePalette, mixinColor } from './color';
export { KBQ_SANITY_CHECKS, KbqCommonModule } from './common-module';
export { CanDisable, CanDisableCtor, mixinDisabled } from './disabled';
export { CanUpdateErrorState, CanUpdateErrorStateCtor, mixinErrorState } from './error-state';
export { HasTabIndex, HasTabIndexCtor, mixinTabIndex } from './tabindex';

export const KBQ_PARENT_ANIMATION_COMPONENT = new InjectionToken<any>('kbq-parent-animation-component');
