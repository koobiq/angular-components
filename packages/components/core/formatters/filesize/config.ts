import { InjectionToken, Provider } from '@angular/core';
import { enUSFormattersData } from '../../locales';

/**
 * @deprecated Will be removed in next major release. Use `KbqMeasurementSystem` instead.
 * @docs-private
 */
export enum MeasurementSystem {
    SI = 'SI',
    IEC = 'IEC'
}

/**
 * Available unit systems for file size formatting.
 * - SI (Metric): 1 KB = 1000 bytes
 * - IEC (Binary): 1 KiB = 1024 bytes
 */
export enum KbqMeasurementSystem {
    SI = 'SI',
    IEC = 'IEC'
}

/** Unit systems as union type */
export type KbqMeasurementSystemType = keyof typeof KbqMeasurementSystem;

export interface KbqUnitSystem {
    abbreviations: string[];
    base: number;
    power: number;
}

/**
 * @deprecated Will be removed in next major release. Use `KbqSizeUnitsConfig` instead.
 * @docs-private
 */
export interface SizeUnitsConfig {
    defaultUnitSystem: string;
    defaultPrecision: number;
    unitSystems: {
        [MeasurementSystem.SI]: KbqUnitSystem;
        [MeasurementSystem.IEC]: KbqUnitSystem;
    };
}

/**
 * Configuration for file size formatting options.
 * Defines the default unit system, precision, and available unit systems.
 */
export interface KbqSizeUnitsConfig {
    /**
     * Default unit system to use (e.g., 'SI' or 'IEC').
     * @see KbqMeasurementSystem
     */
    defaultUnitSystem: KbqMeasurementSystemType;
    /**
     * Default number of decimal places to display in formatted output.
     * @example
     * `2` â†’ "1.23 KB"
     */
    defaultPrecision: number;
    /** @see KbqMeasurementSystem */
    unitSystems: {
        [KbqMeasurementSystem.SI]: KbqUnitSystem;
        [KbqMeasurementSystem.IEC]: KbqUnitSystem;
    };
}

export const KBQ_SIZE_UNITS_DEFAULT_CONFIG: KbqSizeUnitsConfig = enUSFormattersData.sizeUnits;

/**
 * Configuration for converting sizes in different unit systems.
 */
export const KBQ_SIZE_UNITS_CONFIG = new InjectionToken<KbqSizeUnitsConfig>('KbqSizeUnitsConfig');

/** Utility provider for `KBQ_SIZE_UNITS_CONFIG`. */
export const kbqFilesizeFormatterConfigurationProvider = (configuration: Partial<KbqSizeUnitsConfig>): Provider => ({
    provide: KBQ_SIZE_UNITS_CONFIG,
    useValue: { ...KBQ_SIZE_UNITS_DEFAULT_CONFIG, ...configuration }
});
