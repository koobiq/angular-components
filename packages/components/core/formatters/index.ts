import { NgModule } from '@angular/core';
import { DateAdapter, KBQ_DATE_LOCALE } from '../datetime';
import { DateFormatter } from './date/formatter';
import {
    AbsoluteDateFormatterPipe,
    AbsoluteDateShortFormatterPipe,
    AbsoluteDateTimeFormatterPipe,
    AbsoluteShortDateTimeFormatterPipe,
    RangeDateFormatterPipe,
    RangeDateTimeFormatterPipe,
    RangeMiddleDateTimeFormatterPipe,
    RangeShortDateFormatterPipe,
    RangeShortDateTimeFormatterPipe,
    RelativeDateFormatterPipe,
    RelativeDateTimeFormatterPipe,
    RelativeShortDateFormatterPipe,
    RelativeShortDateTimeFormatterPipe
} from './date/formatter.pipe';
import { KbqDecimalPipe, KbqRoundDecimalPipe, KbqTableNumberPipe } from './number/formatter';

@NgModule({
    declarations: [
        KbqDecimalPipe,
        KbqRoundDecimalPipe,
        KbqTableNumberPipe,
        AbsoluteDateFormatterPipe,
        AbsoluteDateTimeFormatterPipe,
        AbsoluteDateShortFormatterPipe,
        AbsoluteShortDateTimeFormatterPipe,
        RelativeDateFormatterPipe,
        RelativeDateTimeFormatterPipe,
        RelativeShortDateFormatterPipe,
        RelativeShortDateTimeFormatterPipe,
        RangeDateFormatterPipe,
        RangeShortDateFormatterPipe,
        RangeDateTimeFormatterPipe,
        RangeShortDateTimeFormatterPipe,
        RangeMiddleDateTimeFormatterPipe
    ],
    exports: [
        KbqDecimalPipe,
        KbqRoundDecimalPipe,
        KbqTableNumberPipe,
        AbsoluteDateFormatterPipe,
        AbsoluteDateTimeFormatterPipe,
        AbsoluteDateShortFormatterPipe,
        AbsoluteShortDateTimeFormatterPipe,
        RelativeDateFormatterPipe,
        RelativeDateTimeFormatterPipe,
        RelativeShortDateFormatterPipe,
        RelativeShortDateTimeFormatterPipe,
        RangeDateFormatterPipe,
        RangeShortDateFormatterPipe,
        RangeDateTimeFormatterPipe,
        RangeShortDateTimeFormatterPipe,
        RangeMiddleDateTimeFormatterPipe
    ],
    providers: [{ provide: DateFormatter, deps: [DateAdapter, KBQ_DATE_LOCALE] }]
})
export class KbqFormattersModule {}

export * from './date/formatter';
export * from './date/formatter.pipe';
export * from './number/formatter';
