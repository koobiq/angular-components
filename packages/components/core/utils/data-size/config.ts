import { InjectionToken } from '@angular/core';
import { enUSFormattersData } from '../../locales';


export enum MeasurementSystem {
    SI = 'SI',
    IEC = 'IEC'
}

export interface SizeUnitsConfig {
    defaultUnitSystem: string,
    defaultPrecision: number,
    unitSystems: {
        [MeasurementSystem.SI]: {
            abbreviations: string[],
            base: number,
            power: number
        },
        [MeasurementSystem.IEC]: {
            abbreviations: string[],
            base: number,
            power: number
        }
    }
}

export const KBQ_SIZE_UNITS_DEFAULT_CONFIG = enUSFormattersData.sizeUnits;

export const KBQ_SIZE_UNITS_CONFIG = new InjectionToken<SizeUnitsConfig>('KbqSizeUnitsConfig');
