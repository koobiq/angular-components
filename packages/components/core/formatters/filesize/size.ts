import { KbqUnitSystem } from './config';

export const KBQ_INVALID_VALUE_ERROR = 'Argument "value" must be a finite number!';

/**
 * @deprecated Will be removed in next major release. Use `getFormattedSizeParts` instead.
 * @docs-private
 */
export const formatDataSize = (
    value: number,
    precision: number,
    system: KbqUnitSystem
): { value: string; unit: string } => {
    const { result, unit } = getHumanizedBytes(value, system);
    let volume: string;

    if (system.abbreviations[0] === unit) {
        volume = result.toString();
    } else {
        volume = result.toFixed(precision).replace(/\./g, ',');
    }

    return {
        value: volume,
        unit
    };
};

/**
 * Converts a byte value into locale-independent file size parts: numeric value and unit abbreviation.
 *
 * @param value - size in bytes.
 * @param system - unit system defining abbreviations and base scaling (SI/IEC).
 * @returns Object with the formatted size info.
 *
 * @example
 * getFormattedSizeParts(1500, 'SI'); // { value: "1.50", unit: "KB" }
 */
export function getFormattedSizeParts(value: number, system: KbqUnitSystem): { value: string; unit: string };
/**
 * Converts a byte value into locale-independent file size parts: numeric value and unit abbreviation.
 *
 * @param value - size in bytes.
 * @param _precision param is deprecated, use `Intl.NumberFormat` rounding options instead.
 * @param system - unit system defining abbreviations and base scaling (SI/IEC).
 * @returns Object with the formatted size info.
 * @deprecated `_precision` - this param will be remove in next major version.
 * @TODO remove deprecated (#DS-3849).
 *
 * @example
 * getFormattedSizeParts(1500, 2, 'SI'); // { value: "1.50", unit: "KB" }
 */
export function getFormattedSizeParts(
    value: number,
    _precision: number,
    system: KbqUnitSystem
): { value: string; unit: string };
export function getFormattedSizeParts(
    value: number,
    _precision: number | KbqUnitSystem,
    system?: KbqUnitSystem
): { value: string; unit: string } {
    let resolvedSystem: KbqUnitSystem | null = null;

    if (arguments.length === 2 && typeof _precision === 'object') {
        resolvedSystem = _precision satisfies KbqUnitSystem;
    }

    if (arguments.length === 3 && typeof _precision === 'number' && system) {
        resolvedSystem = system;
    }

    if (!resolvedSystem) {
        throw new Error('Unexpected arguments size');
    }

    const { result, unit } = getHumanizedBytes(value, resolvedSystem);

    return {
        value: result.toString(),
        unit
    };
}

/**
 * Converts bytes to Kb, Mb, Gb
 *
 * @param value the number of bytes
 * @param system the measurement system
 * @param threshold the lower counting threshold
 */
export const getHumanizedBytes = (
    value: number,
    system: KbqUnitSystem,
    threshold?: number
): { result: number; unit: string } => {
    if (!Number.isFinite(value)) {
        throw new Error(KBQ_INVALID_VALUE_ERROR);
    }

    const calculatedThreshold = Number.isFinite(threshold) ? threshold : Math.pow(system.base, system.power);
    const orderOfMagnitude: number = Math.pow(system.base, system.power);

    let result: number = value;
    let step = 0;
    const len: number = system.abbreviations.length - 1;

    while (step < len) {
        if (calculatedThreshold !== undefined && result < calculatedThreshold) {
            break;
        }

        step++;
        result = value / Math.pow(orderOfMagnitude, step);
    }

    return {
        result,
        unit: system.abbreviations[step]
    };
};
