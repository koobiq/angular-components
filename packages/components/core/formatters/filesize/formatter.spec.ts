import { inject, TestBed } from '@angular/core/testing';
import {
    KBQ_INVALID_VALUE_ERROR,
    KBQ_LOCALE_ID,
    KBQ_SIZE_UNITS_CONFIG,
    KbqDataSizePipe,
    KbqDecimalPipe,
    KbqFormattersModule,
    KbqLocaleService,
    KbqLocaleServiceModule,
    KbqSizeUnitsConfig,
    ruRUFormattersData
} from '@koobiq/components/core';
import { KBQ_SIZE_UNITS_DEFAULT_CONFIG, KbqMeasurementSystem } from './config';
import { formatDataSize, getFormattedSizeParts, getHumanizedBytes } from './size';

describe('Filesize formatter', () => {
    describe(getHumanizedBytes.name, () => {
        it('SI unit system', () => {
            const raw = 1000000; // 1*1000*1000
            const expectedResult = 1;
            const expectedUnit = 'MB';

            const { result, unit } = getHumanizedBytes(
                raw,
                KBQ_SIZE_UNITS_DEFAULT_CONFIG.unitSystems[KbqMeasurementSystem.SI]
            );

            expect(result).toBe(expectedResult);
            expect(unit).toBe(expectedUnit);
        });

        it('IEC unit system', () => {
            const raw = 1048576; // 1*1024*1024
            const expectedResult = 1;
            const expectedUnit = 'MiB';

            const { result, unit } = getHumanizedBytes(
                raw,
                KBQ_SIZE_UNITS_DEFAULT_CONFIG.unitSystems[KbqMeasurementSystem.IEC]
            );

            expect(result).toBe(expectedResult);
            expect(unit).toBe(expectedUnit);
        });

        it('with threshold', () => {
            const raw = 1048576; // 1*1024*1024
            const expectedResult = 1024; // 1024 KiB
            const threshold = 10000; // 1024 KiB
            const expectedUnit = 'KiB';

            const { result, unit } = getHumanizedBytes(
                raw,
                KBQ_SIZE_UNITS_DEFAULT_CONFIG.unitSystems[KbqMeasurementSystem.IEC],
                threshold
            );

            expect(result).toBe(expectedResult);
            expect(unit).toBe(expectedUnit);
        });
    });

    it(formatDataSize.name, () => {
        const raw = 53094588; // 50.63 MiB
        const precision = 2;
        const expected = '50,63';

        const { value } = formatDataSize(
            raw,
            precision,
            KBQ_SIZE_UNITS_DEFAULT_CONFIG.unitSystems[KbqMeasurementSystem.IEC]
        );

        expect(value).toBe(expected);
    });

    describe(getFormattedSizeParts.name, () => {
        const raw = 53094588; // 50.63 MiB
        const selectedUnitSystem = KBQ_SIZE_UNITS_DEFAULT_CONFIG.unitSystems[KbqMeasurementSystem.IEC];

        it('should format value to locale-independent numeric string', () => {
            const { value } = getFormattedSizeParts(raw, selectedUnitSystem);

            expect(value).toMatchSnapshot();
        });

        it('should format value to selected unit system', () => {
            const selectedUnitSystem = KBQ_SIZE_UNITS_DEFAULT_CONFIG.unitSystems[KbqMeasurementSystem.SI];

            const { value } = getFormattedSizeParts(raw, selectedUnitSystem);

            expect(value).toMatchSnapshot();
        });
    });

    describe(KbqDataSizePipe.name, () => {
        let pipe: KbqDataSizePipe;
        let localeService: KbqLocaleService;

        describe('core', () => {
            beforeEach(() => {
                TestBed.configureTestingModule({
                    imports: [KbqFormattersModule, KbqLocaleServiceModule],
                    providers: [
                        KbqDataSizePipe,
                        { provide: KBQ_LOCALE_ID, useValue: 'en-US' }]
                }).compileComponents();
            });

            beforeEach(inject([KbqDataSizePipe, KbqLocaleService], (p: KbqDataSizePipe, l: KbqLocaleService) => {
                pipe = p;
                localeService = l;
            }));

            it('should transform bytes using default config and locale', () => {
                const result = pipe.transform(1500);

                expect(result).toContain('1.5');
                expect(result).toContain(
                    KBQ_SIZE_UNITS_DEFAULT_CONFIG.unitSystems[KBQ_SIZE_UNITS_DEFAULT_CONFIG.defaultUnitSystem]
                        .abbreviations[1]
                );
            });

            it('should throw error for null source', () => {
                const wrapper = () => pipe.transform(null as any);

                expect(wrapper).toThrow(KBQ_INVALID_VALUE_ERROR);
            });

            it('should apply a specific locale if passed', () => {
                const selectedLocale = 'ru-RU';
                const result = pipe.transform(1500, 2, KbqMeasurementSystem.SI, 'ru-RU');

                expect(result).toContain('1,5');
                expect(result).toContain(
                    ruRUFormattersData.sizeUnits.unitSystems[KbqMeasurementSystem.SI].abbreviations[1]
                );
                expect(localeService.id).not.toEqual(selectedLocale);
            });
        });

        describe('with localeService is not provided', () => {
            beforeEach(() => {
                TestBed.configureTestingModule({
                    imports: [KbqFormattersModule],
                    providers: [KbqDataSizePipe]
                }).compileComponents();
            });

            beforeEach(inject([KbqDataSizePipe], (p: KbqDataSizePipe) => (pipe = p)));

            it('should fallback to default config if localeService not provided', () => {
                const result = pipe.transform(1500);

                expect(result).toContain('1,5');
                expect(result).toContain(
                    KBQ_SIZE_UNITS_DEFAULT_CONFIG.unitSystems[KBQ_SIZE_UNITS_DEFAULT_CONFIG.defaultUnitSystem]
                        .abbreviations[1]
                );
            });
        });

        describe('with externalConfig provided', () => {
            const externalConfig: KbqSizeUnitsConfig = {
                defaultUnitSystem: KbqMeasurementSystem.SI,
                defaultPrecision: 3,
                unitSystems: {
                    SI: {
                        abbreviations: ['BTEST', 'KBTEST', 'MBTEST', 'GBTEST', 'TBTEST'],
                        base: 10,
                        power: 3
                    },
                    IEC: {
                        abbreviations: ['BTEST', 'KiBTEST', 'MiBTEST', 'GiBTEST', 'TiBTEST'],
                        base: 2,
                        power: 10
                    }
                }
            };

            beforeEach(() => {
                TestBed.configureTestingModule({
                    imports: [KbqFormattersModule],
                    providers: [
                        KbqDataSizePipe,
                        {
                            provide: KBQ_SIZE_UNITS_CONFIG,
                            useValue: externalConfig
                        }
                    ]
                }).compileComponents();
            });

            beforeEach(inject([KbqDataSizePipe], (p: KbqDataSizePipe) => (pipe = p)));

            it('should prioritize external config over localeService', () => {
                const result = pipe.transform(1500);
                const resAbbreviation = externalConfig.unitSystems[externalConfig.defaultUnitSystem].abbreviations[1];
                const localizedConfig: KbqSizeUnitsConfig = localeService.getParams('sizeUnits');

                expect(result).toContain('1,5');
                expect(result).toContain(resAbbreviation);

                expect(resAbbreviation).not.toEqual(
                    localizedConfig.unitSystems[localizedConfig.defaultUnitSystem].abbreviations[1]
                );
            });
        });

        describe('with numberPipe is not provided', () => {
            beforeEach(() => {
                TestBed.configureTestingModule({
                    providers: [
                        KbqDataSizePipe,
                        { provide: KbqDecimalPipe, useValue: null }]
                }).compileComponents();
            });

            beforeEach(inject([KbqDataSizePipe], (p: KbqDataSizePipe) => (pipe = p)));

            it('should fallback to default config if localeService not provided', () => {
                const result = pipe.transform(1500);
                const { value, unit } = getFormattedSizeParts(
                    1500,
                    KBQ_SIZE_UNITS_DEFAULT_CONFIG.unitSystems[KbqMeasurementSystem.SI]
                );

                expect(result).toContain(value);
                expect(result).toContain(unit);
            });
        });
    });
});
