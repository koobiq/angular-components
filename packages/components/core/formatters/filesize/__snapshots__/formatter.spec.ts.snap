// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Filesize formatter getFormattedSizeParts should format value to locale-independent numeric string 1`] = `"50.634944915771484"`;

exports[`Filesize formatter getFormattedSizeParts should format value to selected unit system 1`] = `"53.094588"`;
